File: exams/u3/saraguro/EXAMEN_3P_Strategy_Pattern/SortingApp/src/strategies/QuickSort.js
class QuickSort {
    sort(array) {
        if (array.length <= 1) return array;
        let pivot = array[array.length - 1];
        let left = [];
        let right = [];
        for (let i = 0; i < array.length - 1; i++) {
            if (array[i] < pivot) left.push(array[i]);
            else right.push(array[i]);
        }
        return [...this.sort(left), pivot, ...this.sort(right)];
    }
}

module.exports = QuickSort;

File: hw/mera/u3/hw 30/StrategyFinalExam/src/main/java/com/jsons/strategyfinalexam/controller/QuickSort.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.jsons.strategyfinalexam.controller;

/**
 *
 * @author Ruben Benavides, Jsons, DCCO-ESPE
 */
public class QuickSort extends SortingStrategy{

    @Override
    public int[] sort(int[] nums){
        nums = sort(nums, 0, nums.length-1);
        return nums;
    }
    
    public int[] sort(int[] nums, int low, int high) {
        if (low < high) {
            int pi = partition(nums, low, high);
            sort(nums, low, pi - 1);
            sort(nums, pi + 1, high);
        }
        
        return nums;
    }
    
    public static int partition(int[] nums, int low, int high) {
        int pivot = nums[high];
        int i = low - 1;
        
        for (int j = low; j < high; j++) {
            if (nums[j] < pivot) {
                i++;
                int temp = nums[i];
                nums[i] = nums[j];
                nums[j] = temp;
            }
        }
        
        int temp = nums[i + 1];
        nums[i + 1] = nums[high];
        nums[high] = temp;
        
        return i + 1;
    }

}
